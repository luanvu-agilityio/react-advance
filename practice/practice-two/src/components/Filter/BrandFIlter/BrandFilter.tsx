'use client'

import { CheckIcon } from 'lucide-react'
import {
  SectionTitle,
  BrandSection,
  BrandItems,
  CheckboxContainer,
  CheckboxRoot,
  CheckboxIndicator,
  Label,
} from '../FilterStyles'
import type { BrandProps } from 'types/Filter'

interface BrandFilterProps {
  brands: BrandProps[]
  onBrandChange: (index: number) => void
}

export const BrandFilter = ({ brands, onBrandChange }: BrandFilterProps) => {
  if (brands.length === 0) return null

  return (
    <BrandSection>
      <SectionTitle>Brands</SectionTitle>
      <BrandItems>
        {brands.map((brand, index) => (
          <CheckboxContainer key={brand.name}>
            <CheckboxRoot
              checked={brand.selected}
              onCheckedChange={() => onBrandChange(index)}
              id={`brand-${index}`}
            >
              <CheckboxIndicator>
                <CheckIcon strokeWidth={2} style={{ paddingTop: '5px' }} />
              </CheckboxIndicator>
            </CheckboxRoot>
            <Label htmlFor={`brand-${index}`}>{brand.name}</Label>
          </CheckboxContainer>
        ))}
      </BrandItems>
    </BrandSection>
  )
}
