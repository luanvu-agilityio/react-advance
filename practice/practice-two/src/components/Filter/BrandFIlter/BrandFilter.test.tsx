import { render, screen, fireEvent } from '@testing-library/react'
import { BrandFilter } from './BrandFilter'
import type { BrandProps } from 'types/Filter'

describe('BrandFilter Component', () => {
  const mockBrands: BrandProps[] = [
    { name: 'United Farms', selected: false },
    { name: 'Organic Farms', selected: true },
    { name: 'Fresh Fields', selected: false },
  ]

  const mockOnBrandChange = jest.fn()

  beforeEach(() => {
    mockOnBrandChange.mockClear()
  })

  // SNAPSHOT TESTS FOR RENDERING
  describe('Rendering', () => {
    it('renders different states correctly', () => {
      // Default state with mixed selections
      const { container: defaultContainer } = render(
        <BrandFilter brands={mockBrands} onBrandChange={mockOnBrandChange} />
      )
      expect(defaultContainer).toMatchSnapshot('mixed-selections')

      // All selected state
      const allSelected: BrandProps[] = mockBrands.map((brand) => ({
        ...brand,
        selected: true,
      }))
      const { container: allSelectedContainer } = render(
        <BrandFilter brands={allSelected} onBrandChange={mockOnBrandChange} />
      )
      expect(allSelectedContainer).toMatchSnapshot('all-selected')

      // None selected state
      const noneSelected: BrandProps[] = mockBrands.map((brand) => ({
        ...brand,
        selected: false,
      }))
      const { container: noneSelectedContainer } = render(
        <BrandFilter brands={noneSelected} onBrandChange={mockOnBrandChange} />
      )
      expect(noneSelectedContainer).toMatchSnapshot('none-selected')

      // Empty brands array
      const { container: emptyContainer } = render(
        <BrandFilter brands={[]} onBrandChange={mockOnBrandChange} />
      )
      expect(emptyContainer).toMatchSnapshot('empty-brands')
    })
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Interactive behavior', () => {
    it('calls onBrandChange with correct index when checkbox is clicked', () => {
      render(
        <BrandFilter brands={mockBrands} onBrandChange={mockOnBrandChange} />
      )

      // Click on the first checkbox label
      const firstBrandLabel = screen.getByText('United Farms')
      fireEvent.click(firstBrandLabel)

      // Check if onBrandChange was called with the correct index
      expect(mockOnBrandChange).toHaveBeenCalledWith(0)
    })

    it('handles clicking on already selected checkbox', () => {
      render(
        <BrandFilter brands={mockBrands} onBrandChange={mockOnBrandChange} />
      )

      // Click on the second checkbox which is already selected
      const secondBrandLabel = screen.getByText('Organic Farms')
      fireEvent.click(secondBrandLabel)

      // Check if onBrandChange was called with the correct index
      expect(mockOnBrandChange).toHaveBeenCalledWith(1)
    })
  })
})
