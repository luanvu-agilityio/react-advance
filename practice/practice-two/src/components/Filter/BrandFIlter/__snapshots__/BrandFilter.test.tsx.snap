// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrandFilter Component Rendering renders different states correctly: all-selected 1`] = `
<div>
  <div
    class="sc-kpOvIu dzcghV"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Brands
    </h2>
    <div
      class="sc-icnseD fNZWKN"
    >
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="true"
          class="sc-cXawGu VtGFV"
          data-state="checked"
          id="brand-0"
          role="checkbox"
          type="button"
          value="on"
        >
          <span
            class="sc-jJAtPt dVvwCd"
            data-state="checked"
            style="pointer-events: none;"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-check"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="padding-top: 5px;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 6 9 17l-5-5"
              />
            </svg>
          </span>
        </button>
        <label
          class="sc-cUiCeM limFCd"
          for="brand-0"
        >
          United Farms
        </label>
      </div>
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="true"
          class="sc-cXawGu VtGFV"
          data-state="checked"
          id="brand-1"
          role="checkbox"
          type="button"
          value="on"
        >
          <span
            class="sc-jJAtPt dVvwCd"
            data-state="checked"
            style="pointer-events: none;"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-check"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="padding-top: 5px;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 6 9 17l-5-5"
              />
            </svg>
          </span>
        </button>
        <label
          class="sc-cUiCeM limFCd"
          for="brand-1"
        >
          Organic Farms
        </label>
      </div>
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="true"
          class="sc-cXawGu VtGFV"
          data-state="checked"
          id="brand-2"
          role="checkbox"
          type="button"
          value="on"
        >
          <span
            class="sc-jJAtPt dVvwCd"
            data-state="checked"
            style="pointer-events: none;"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-check"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="padding-top: 5px;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 6 9 17l-5-5"
              />
            </svg>
          </span>
        </button>
        <label
          class="sc-cUiCeM limFCd"
          for="brand-2"
        >
          Fresh Fields
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`BrandFilter Component Rendering renders different states correctly: empty-brands 1`] = `<div />`;

exports[`BrandFilter Component Rendering renders different states correctly: mixed-selections 1`] = `
<div>
  <div
    class="sc-kpOvIu dzcghV"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Brands
    </h2>
    <div
      class="sc-icnseD fNZWKN"
    >
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="false"
          class="sc-cXawGu VtGFV"
          data-state="unchecked"
          id="brand-0"
          role="checkbox"
          type="button"
          value="on"
        />
        <label
          class="sc-cUiCeM limFCd"
          for="brand-0"
        >
          United Farms
        </label>
      </div>
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="true"
          class="sc-cXawGu VtGFV"
          data-state="checked"
          id="brand-1"
          role="checkbox"
          type="button"
          value="on"
        >
          <span
            class="sc-jJAtPt dVvwCd"
            data-state="checked"
            style="pointer-events: none;"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-check"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="padding-top: 5px;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 6 9 17l-5-5"
              />
            </svg>
          </span>
        </button>
        <label
          class="sc-cUiCeM limFCd"
          for="brand-1"
        >
          Organic Farms
        </label>
      </div>
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="false"
          class="sc-cXawGu VtGFV"
          data-state="unchecked"
          id="brand-2"
          role="checkbox"
          type="button"
          value="on"
        />
        <label
          class="sc-cUiCeM limFCd"
          for="brand-2"
        >
          Fresh Fields
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`BrandFilter Component Rendering renders different states correctly: none-selected 1`] = `
<div>
  <div
    class="sc-kpOvIu dzcghV"
  >
    <h2
      class="sc-fYsHOw gpnbOD"
    >
      Brands
    </h2>
    <div
      class="sc-icnseD fNZWKN"
    >
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="false"
          class="sc-cXawGu VtGFV"
          data-state="unchecked"
          id="brand-0"
          role="checkbox"
          type="button"
          value="on"
        />
        <label
          class="sc-cUiCeM limFCd"
          for="brand-0"
        >
          United Farms
        </label>
      </div>
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="false"
          class="sc-cXawGu VtGFV"
          data-state="unchecked"
          id="brand-1"
          role="checkbox"
          type="button"
          value="on"
        />
        <label
          class="sc-cUiCeM limFCd"
          for="brand-1"
        >
          Organic Farms
        </label>
      </div>
      <div
        class="sc-jMsorb cyLRQD"
      >
        <button
          aria-checked="false"
          class="sc-cXawGu VtGFV"
          data-state="unchecked"
          id="brand-2"
          role="checkbox"
          type="button"
          value="on"
        />
        <label
          class="sc-cUiCeM limFCd"
          for="brand-2"
        >
          Fresh Fields
        </label>
      </div>
    </div>
  </div>
</div>
`;
